### PR TITLE
Simplify iterator logic for Fuse

### PR DESCRIPTION
**Pull request by frewsxcv**

### CODE REVIEW

**Comment by bluss** [comment]
It doesn't simplify, it just writes the same thing in a shorter way. Does it matter that we use an extra closure here that the optimizier needs to re-inline? Because of the likely code size increase (in code given to llvm), I think this is a net negative unfortunately. Sorry to be negative about your minor fix, I'm just trying to weigh the change's merits.

**Comment by frewsxcv** [comment]
> It doesn't simplify, it just writes the same thing in a shorter way.

I think it is cleaner and more idiomatic. I am utilizing `Option::or_else` in the way it was intended to be used. 

>  Because of the likely code size increase

Compare the LLVM IR for these two examples:

* http://is.gd/QHJMWt
* http://is.gd/uTwlgi

They're identical as far as I know, and one of them uses a closure, so I'm not convinced code will increase.

**Comment by lfairy** [comment]
As scott mentioned on IRC, a better rewrite would be:

```rust
let x = self.iter.next();
if x.is_none() {
    self.done = true;
}
x
```

as that's much easier to understand.

**Comment by frewsxcv** [comment]
@lfairy Good idea, just force pushed with that change

File: src/libcore/iter.rs [hunk 1]
```diff
@@ -2224,13 +2222,11 @@ impl<I> DoubleEndedIterator for Fuse<I> where I: DoubleEndedIterator {
         if self.done {
             None
         } else {
-            match self.iter.next_back() {
-                None => {
-                    self.done = true;
-                    None
-                }
-                x => x
+            let next = self.iter.next();
```
Comment (Gankro) [inline src/libcore/iter.rs]: Uhhhh next_back...?

Comment (frewsxcv) [inline src/libcore/iter.rs]: Good catch, fixed

File: src/libcore/iter.rs [hunk 2]
```diff
@@ -2224,13 +2222,11 @@ impl<I> DoubleEndedIterator for Fuse<I> where I: DoubleEndedIterator {
...
         } else {
-            match self.iter.next_back() {
-                None => {
-                    self.done = true;
-                    None
-                }
-                x => x
+            let next = self.iter.next_back();
+            if next.is_none() {
+                self.done = true;
```
Comment (Gankro) [inline src/libcore/iter.rs]: Hmm, what about just `self.done = next.is_none()`? No branch at all, but unconditional mutation.

Comment (lfairy) [inline src/libcore/iter.rs]: I think writing `self.done = true` makes explicit the transition from `false` -> `true`.

But I'm not too attached to the idea.

Comment (frewsxcv) [inline src/libcore/iter.rs]: I was about to say the same thing as @lfairy , but I too don't feel too strongly about it, especially considering how small of a change this commit is

Comment (Gankro) [inline src/libcore/iter.rs]: Anytime I see a `done` variable I instantly go to "this is a one-transition boolean", personally.

Comment (frewsxcv) [inline src/libcore/iter.rs]: Force pushed with that change

**Comment by bluss** [comment]
If the IR is identical, it's because it's after optimization, I was thinking of the size of llvm's input from rustc from the beginning. Now with the closure gone, the point is moot and that's well.

**Comment by cristicbz** [comment]
I think this change breaks the fuse() impl. Before done would only be set to true if next() returned None. Now it would be set to false again if it isn't.

You want the rewrite @lfairy posted with the self.done inside the if.

**Comment by frewsxcv** [comment]
> Now it would be set to false again if it isn't.

@cristicbz Can you give me an example where it would be set to `false` again?

**Comment by cristicbz** [comment]
Never mind, sorry, I didn't read it properly---I missed the big if self.done.

**Pull request merged**